### PR TITLE
build-engine API changes from "entry" to "feature"-based

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: cocos-creator/3d-tasks#

Changelog:
 *

### CODE REVIEW

File: cc.config.json [hunk 1]
```diff
@@ -1,12 +1,45 @@
...
+        "terrain": { "modules": [ "terrain" ] },
+        "webview": { "modules": [ "webview" ] },
+        "tween": { "modules": [ "tween" ] },
+        "tiled-map": { "modules": [ "tiled-map" ] },
+        "spine": { "modules": [ "spine" ] },
+        "dragonbones": { "modules": [ "dragonbones" ] }
+    },
+    "index": {
+        "modules": {
+            "spine": { "ns": "sp" }
```
Comment (shrinktofit) [inline cc.config.json:25]: @PatriceJiang

File: cc.config.json [hunk 2]
```diff
@@ -1,12 +1,45 @@
...
+        }
+    },
+    "moduleOverrides": [
+        {
+            "test": "context.mode === 'BUILD'",
```
Comment (shrinktofit) [inline cc.config.json:30]: @jareguo

File: cc.config.json [hunk 3]
```diff
@@ -1,12 +1,45 @@
...
+            "overrides": {
+                "cocos/core/data/deserialize-dynamic.ts": "cocos/core/data/deserialize-dynamic-empty.ts"
+            }
+        },
+        {
+            "test": "context.platform === 'NATIVE'",
```
Comment (shrinktofit) [inline cc.config.json:36]: @minggo

File: cocos/spine/skeleton.ts [hunk 4]
```diff
@@ -14,7 +14,6 @@ import { UI } from '../core/renderer/ui/ui';
 import { Graphics } from '../ui/components/graphics';
 import { MaterialInstance } from '../core/renderer';
 import { js } from '../core/utils/js';
-import { sp } from '../../exports/spine';
```
Comment (shrinktofit) [inline cocos/spine/skeleton.ts:17]: @PatriceJiang

File: exports/spine.ts [hunk 5]
```diff
@@ -1,3 +1 @@
-
-import * as sp from '../cocos/spine';
-export {sp};
\ No newline at end of file
+export * from '../cocos/spine';
```
Comment (shrinktofit) [inline exports/spine.ts:1]: @PatriceJiang

File: index.ts [hunk 6]
```diff
@@ -21,5 +21,6 @@ export * from './exports/physics-2d-framework';
 export * from './exports/physics-2d-box2d';
 export * from './exports/particle-2d';
 export * from './exports/tiled-map';
-export * from './exports/spine';
+import * as sp from './exports/spine';
```
Comment (shrinktofit) [inline index.ts:24]: @PatriceJiang

File: scripts/build-engine/src/stats-query.ts [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+interface IndexConfig {
+    modules?: Record<string, {
+        /**
+         * If specified, export contents of the module into a namespace specified by `ns`
+         * and then export that namespace into `'cc'`.
+         * If not specified, contents of the module will be directly exported into `'cc'`.
+         */
+        ns?: string;
```
Comment (shrinktofit) [inline scripts/build-engine/src/stats-query.ts:31]: @PatriceJiang

**Review by PatriceJiang — APPROVED** [review]

File: cc.config.json (hunk 3 above)
Comment (minggo) [reply cc.config.json:36]: It is ok to me if it can replace the .ts files correctly.

**Review by yanOO1497 — APPROVED** [review]

**Pull request merged**